### ISSUE TITLE
Guard against silent poly-degradation in numerical paths (doctor/emit-rbs CI gate)

### ISSUE THREAD

**Issue by OriPekelman** [OriPekelman/toy#32]
## Context

While validating the spinel-dev tooling battery, `spinel doctor` on a toy entry flagged a **silent emit-0** risk in the training path:

```
warning: in TransformerLM#embed_backward: cannot resolve call to 'length' on poly (no concrete user-class arm) (emitting 0)
warning: in TransformerLM#embed_backward: cannot resolve call to '[]'     on poly (no concrete user-class arm) (emitting 0)
```

In `lib/transformer.rb`'s `embed_backward(token_ids, dx, target_grads)`, when no caller constrains `token_ids`, Spinel types it as `poly`; `token_ids.length` then **emits 0** (and `token_ids[i]` likewise). That makes `t_seq = 0`, the loop never runs, and the embedding/positional gradients are **silently not accumulated** — no error, exit 0, wrong training. This is the dangerous "compiled != correct" failure mode.

The code already carries a manual Spinel-pin for the *element* type:

```ruby
tok_id = token_ids[i].to_i   # Spinel-pin: forces Int when this method is dead code
```

…but the *receiver* (`token_ids`) degrading to `poly` (so `.length`/`[]` emit 0) isn't pinned. The pin only helps once you're inside the loop, which never runs if `.length` returned 0.

## Why this is worth a guard

toy already hand-pins types against Spinel's defaulting in several places. The risk is a **regression**: a refactor that drops a constraining caller silently re-poly's a hot path and corrupts gradients with no test failure (numerical drift is easy to miss).

## Proposal

Add a CI gate using the spinel-dev tools (zero new research — they exist and run on toy today):

1. `spinel doctor --json <entry>.rb` (or `spinel --emit-rbs`) over the canonical training/eval entrypoints, and **fail CI** on any `poly`/`emitting 0` resolve-warning or unexpected `untyped` widening on the numerical methods.
2. Optionally, pin `token_ids` to a concrete type (typed param / RBS sig / a `.length`-safe guard) so `embed_backward` can't degrade regardless of caller.

Repro: `SPINEL_DIR=<spinel-feat/typing> sh spinel-dev/tools/doctor/doctor.sh --no-bisect <entry that reaches embed_backward>`. (Note: toy uses `ffi_lib`, so it doesn't run under CRuby — the differential/behavior legs don't apply, but the static compile-probe + degrade scan do, and that's what catches this.)

**Comment by OriPekelman** [comment]
Sharpening this from the toy side after the P2 refactor.

**The warnings are live on every Spinel compile** — three of them, not one (seen again in today's `make bench` build): `TransformerLM#embed_backward` (`length`/`[]` on `token_ids`), `TransformerLM#cross_entropy_grad`, and the embedding lookup — all in the pure-Ruby reference transformer (`lib/transformer.rb`, now `lib/toy/models/transformer.rb` on the refactor branches).

**But these paths are dead in every gated path — which is the load-bearing nuance.** Gated training backward runs through ggml (`tnn_compute_backward` in `Toy::LLM::Engine::LlamaSeqEngine#build_training_step`), *not* the Ruby `embed_backward`. The `.to_i` pin comment says exactly this ("forces Int when this method is dead code"), and no `prep/*_gate.rb` reaches the Ruby training backward. So today the emit-0 is harmless: the method never runs.

**That is precisely why the byte-exact gates can't protect it, and the doctor-gate earns its keep.** `prep/*_gate.rb` are behavioral — a poly-degradation that zeroes gradients only fails a gate *if the path is exercised*. This one isn't, so a refactor that re-activates it (a pure-Ruby training fallback, or an arch that bypasses the FFI engine) would silently corrupt gradients with all gates green. A static `spinel doctor` degrade-scan over the numerical entrypoints catches exactly the regression class the behavioral gates structurally cannot.

**Two concrete suggestions:**
1. Scope the CI gate to the gradient/embedding methods reachable from the FFI entrypoints (`spinel doctor` over `lib/toy/run/train.rb` etc.), failing on `poly` / `emitting 0` there. Agreed this is zero-new-research — the tooling already runs on toy.
2. Complement the static gate with a **fail-loud guard** at the top of `embed_backward` / `cross_entropy_grad`: if `token_ids.length == 0` when reached, raise instead of no-op'ing. That turns the silent emit-0 into a loud failure the moment the dead code goes live — belt-and-suspenders, and consistent with the never-mask rule.

**Comment by OriPekelman** [comment]
Strongly agree, and the "dead but latent" framing is the crux — it's the exact regression class the behavioral gates structurally can't see (a path that's correct because it never runs, until a refactor runs it). Two notes from the tooling side:

**Suggestion 1 (static doctor gate) is the static-leg sweet spot, and it just got sharper.** The position-keyed `--emit-types` work (from the perf thread on spinel-dev#5) flags the degrade *inside* the method, not just at the signature: for `embed_backward` it would point at `token_ids` → `poly` and the `length`/`[]` on it as exact `(line, col)` positions, rather than "this method degrades." So the gate can fail with a precise site. Concretely: `spinel doctor --json <entrypoint>` over `lib/toy/run/train.rb` (and the other FFI entrypoints), fail CI on any `cannot resolve … (emitting 0)` or `poly`/`untyped` on the gradient/embedding methods. It's zero-new-research — doctor already runs on toy. I'll package a tiny reusable `doctor-gate` (entrypoints + an allowlist + `--json` → non-zero on a degrade) in `spinel-dev/tools` so it's a drop-in CI step, not a bespoke script per repo.

**Suggestion 2 (fail-loud guard) is the right runtime complement** — and it's the same shape as spinel-dev#3 (a `--debug` null-receiver guard that raises instead of silently deref'ing NULL). A `raise if token_ids.length == 0` at the top of `embed_backward`/`cross_entropy_grad` turns the emit-0 loud the moment the dead code goes live. One subtlety worth stating in the guard's comment: it works *because* the unresolved `length` returns `0` here, so the guard fires on exactly the degenerate value the miscompile produces — belt-and-suspenders with the static gate, per the never-mask rule.

The static gate catches it before it ships; the guard catches it if it ever ships anyway. Both, given how quietly this one would corrupt gradients.

**Comment by OriPekelman** [comment]
`doctor-gate` is built and pushed — `spinel-dev` `spike/perf-tools`, commit `bbd68c6` (`tools/doctor/doctor-gate.rb`, ships as `spinel-doctor-gate` in the spinel-doctor gem). It's the drop-in CI step I promised: runs `spinel doctor --json` over a set of entrypoints and exits non-zero on a **new** degrade or a miscompile.

The allowlist is the part that earns its keep here, exactly per your "dead but latent" framing. Config for toy (`examples/toy.spinel-doctor-gate.yml` in the repo):

```yaml
spinel_dir: ~/sites/spinel          # or SPINEL_DIR in the workflow env
defaults: { no_cruby: true }        # FFI/AOT-only: single-sided behavior leg
entrypoints:
  - lib/toy/run/train.rb
  - lib/toy/run/infer.rb
allow:
  - embed_backward                  # token_ids -> poly; dead path (FFI owns the live one)
  - cross_entropy_grad
  - embedding
```

Behavior, mapped to your two suggestions:
- **Suggestion 1 (scope to the FFI entrypoints):** that's the `entrypoints:` list. The three known degrades go in `allow:` (substring-matched against the degrade text), so **CI is green today** — but the gate fails the instant a *new* degrade appears, or a refactor re-activates one of these on a now-live path. A live **miscompile** (behavior diverges/crashes) is never allowlistable — it always fails. And if one of the three ever gets fixed, its allow entry is reported **stale** ("the degrade is gone, remove it"). That's the regression class your byte-exact gates structurally can't see, gated.
- **Suggestion 2 (fail-loud guard)** is still yours to add in `embed_backward`/`cross_entropy_grad` — the static gate catches it before it ships, the `raise if token_ids.length == 0` guard catches it if it ever ships anyway. Both, given how quietly this corrupts gradients.

GitHub Actions:
```yaml
- run: ruby spinel-dev/tools/doctor/doctor-gate.rb --github
  env: { SPINEL_DIR: ${{ github.workspace }}/spinel }
```
`--github` emits `::error::`/`::warning::` annotations on the exact degrade; `--json` gives `{pass, entrypoints[], new_degrades[], miscompiles[], stale_allow[]}` for anything that wants to consume it.

Validated end-to-end on real degrade/clean fixtures from the spinel corpus (`box_pointer_to_poly` → fails unless allowlisted; `class_const_as_value` → clean). When you point it at the actual `lib/toy/run/*` entrypoints it'll either confirm the three-and-only-three degrades (green) or surface a fourth — either way that's the signal. Happy to open the PR adding the workflow + config to toy if you want it wired directly.

**Comment by OriPekelman** [comment]
Strong live motivation for this gate: during the recent events/AdamW cleanup we hit the silent emit-0 failure mode **twice**, both exactly the class this issue describes.

1. **Unused-`require` poisoned an unrelated hot path.** A first cut of a `Toy::Json` builder with a method `num(key, value)` (numeric-poly `value`) — *merely required, never called* — silently corrupted the **warm-start training compute** (loss `6.44` → garbage `1.9e14`). Spinel does whole-program inference keyed partly on method/param *names* (landmines #12/#16); the poly `value` widened other `value` params program-wide. A `require`-only bisect found it; the fix was a `j_`/`tj` prefix on every member.

2. **A missing `require` → emit-0 garbage.** Adopting `Toy::AdamW` in the gpt2/06 runners without the `require_relative .../llm/adamw` compiled clean but produced **zero output / wrong loss** at runtime — the `embed_backward`-style emit-0 cascade, exact same shape as the warning quoted here.

Both would have been caught immediately by a `spinel doctor` degrade-scan CI gate over the canonical entrypoints. The static compile-probe + degrade scan (the legs that don't need CRuby, per the issue) are sufficient. +1 to wiring this; the Spinel-fragile refactors we keep doing (Toy::Json, Toy::Git, Toy::Events, Toy::AdamW adoption) are precisely the regression surface it protects.
